File: go/datastructures/stack.go
package datastructures

// Stack implements a LIFO stack using a linked list.
// Time: O(1) for all operations
// Space: O(n)
type Stack[T any] struct {
	head *stackNode[T]
	size int
}

type stackNode[T any] struct {
	value T
	next  *stackNode[T]
}

func NewStack[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (s *Stack[T]) Push(value T) {
	s.head = &stackNode[T]{value: value, next: s.head}
	s.size++
}

func (s *Stack[T]) Pop() (T, bool) {
	var zero T
	if s.head == nil {
		return zero, false
	}
	value := s.head.value
	s.head = s.head.next
	s.size--
	return value, true
}

func (s *Stack[T]) Peek() (T, bool) {
	var zero T
	if s.head == nil {
		return zero, false
	}
	return s.head.value, true
}

func (s *Stack[T]) IsEmpty() bool {
	return s.size == 0
}

func (s *Stack[T]) Size() int {
	return s.size
}

func (s *Stack[T]) Clear() {
	s.head = nil
	s.size = 0
}

File: haskell/src/DataStructures/Stack.hs
{-|
Module      : DataStructures.Stack
Description : LIFO stack implementation
Stability   : stable

A purely functional stack implementation.
Time: O(1) for all operations
Space: O(n)
-}
module DataStructures.Stack
  ( Stack
  , empty
  , isEmpty
  , push
  , pop
  , peek
  , size
  , fromList
  , toList
  ) where

-- | A stack represented as a list (head is top)
newtype Stack a = Stack [a]
  deriving (Show, Eq)

-- | Create an empty stack
empty :: Stack a
empty = Stack []

-- | Check if stack is empty
isEmpty :: Stack a -> Bool
isEmpty (Stack xs) = null xs

-- | Push an element onto the stack
push :: a -> Stack a -> Stack a
push x (Stack xs) = Stack (x : xs)

-- | Pop an element from the stack
pop :: Stack a -> Maybe (a, Stack a)
pop (Stack [])     = Nothing
pop (Stack (x:xs)) = Just (x, Stack xs)

-- | Peek at the top element without removing
peek :: Stack a -> Maybe a
peek (Stack [])    = Nothing
peek (Stack (x:_)) = Just x

-- | Get the number of elements in the stack
size :: Stack a -> Int
size (Stack xs) = length xs

-- | Create a stack from a list (first element becomes top)
fromList :: [a] -> Stack a
fromList = Stack

-- | Convert stack to list (top element first)
toList :: Stack a -> [a]
toList (Stack xs) = xs
